File: cli/internal/model/events.go
package model

import (
	"fmt"
	"strings"
	"sync"
	"time"
)

const taskEndGraceS = 2.0 // API 源任务结束事件宽限期：等日志源补全完整统计

// EventDetector 状态迁移检测（与面板 events.EventDetector 行为一致）。
type EventDetector struct {
	mu sync.Mutex

	events []Event
	maxlen int

	llamaOnline *bool
	taskRunning bool
	runningID   *int
	modelPath   *string

	pendingEnd   *taskEndArgs
	pendingTimer *time.Timer
	endedIDs     map[int]bool
	endedOrder   []int

	alertLevels map[string]alertLevel
}

type alertLevel struct {
	level    string
	lastEmit *float64
}

type taskEndArgs struct {
	ts             float64
	taskID         *int
	totalTokens    *int
	durationS      *float64
	avgTPS         *float64
	mtpAcceptance  *float64
	ctxUsed        *int
	note           string
}

func NewEventDetector(maxlen int) *EventDetector {
	if maxlen <= 0 {
		maxlen = 200
	}
	return &EventDetector{
		maxlen:      maxlen,
		endedIDs:    make(map[int]bool),
		alertLevels: make(map[string]alertLevel),
	}
}

func (d *EventDetector) Emit(ts float64, level, typ, msg string) {
	d.mu.Lock()
	defer d.mu.Unlock()
	if ts == 0 {
		ts = float64(time.Now().Unix())
	}
	d.events = append(d.events, Event{TS: ts, Level: level, Type: typ, Msg: msg})
	if len(d.events) > d.maxlen {
		d.events = d.events[len(d.events)-d.maxlen:]
	}
}

func (d *EventDetector) SetLlamaOnline(ts float64, online bool, modelName string) {
	d.mu.Lock()
	defer d.mu.Unlock()
	if d.llamaOnline == nil {
		d.llamaOnline = &online
		if online {
			d.emitLocked(ts, "info", "llama_up", "llama 上线"+modelSuffix(modelName))
		}
		return
	}
	if *d.llamaOnline != online {
		d.llamaOnline = &online
		if online {
			d.emitLocked(ts, "info", "llama_up", "llama 恢复在线"+modelSuffix(modelName))
		} else {
			d.resetTaskStateLocked()
			d.emitLocked(ts, "error", "llama_down", "llama 离线")
		}
	}
}

func (d *EventDetector) ResetTaskState() {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.resetTaskStateLocked()
}

func (d *EventDetector) resetTaskStateLocked() {
	d.taskRunning = false
	d.runningID = nil
	d.endedIDs = make(map[int]bool)
	d.endedOrder = nil
	d.cancelPendingLocked()
}

func (d *EventDetector) SetTaskRunning(ts float64, running bool, taskID *int, promptTokens *int) {
	d.mu.Lock()
	defer d.mu.Unlock()
	if running && !d.taskRunning {
		d.taskRunning = true
		d.runningID = taskID
		d.emitLocked(ts, "info", "task_start", taskStartMsg(taskID, promptTokens))
	} else if running && d.taskRunning && taskID != nil && (d.runningID == nil || *d.runningID != *taskID) {
		d.firePendingLocked()
		d.taskRunning = true
		d.runningID = taskID
		d.emitLocked(ts, "info", "task_start", taskStartMsg(taskID, promptTokens))
	} else if !running && d.taskRunning {
		d.taskRunning = false
		d.runningID = nil
		d.cancelPendingLocked()
		msg := "任务结束"
		if taskID != nil {
			msg = fmt.Sprintf("任务 #%d 结束", *taskID)
		}
		d.emitLocked(ts, "info", "task_end", msg)
	}
}

// TaskEndWithStats 任务结束（带统计）。完整统计（MTP/上下文）立即发；
// 不完整（API 源先到）延迟宽限期给日志源补全，超时用部分统计兜底。
func (d *EventDetector) TaskEndWithStats(ts float64, taskID *int, totalTokens *int,
	durationS, avgTPS, mtpAcceptance *float64, ctxUsed *int, note string) {
	d.mu.Lock()
	defer d.mu.Unlock()
	complete := mtpAcceptance != nil || ctxUsed != nil
	if complete {
		d.cancelPendingLocked()
		d.emitTaskEndLocked(ts, taskID, totalTokens, durationS, avgTPS, mtpAcceptance, ctxUsed, note)
		return
	}
	if !d.taskRunning {
		return
	}
	if d.pendingEnd == nil {
		args := &taskEndArgs{ts: ts, taskID: taskID, totalTokens: totalTokens,
			durationS: durationS, avgTPS: avgTPS, mtpAcceptance: mtpAcceptance, ctxUsed: ctxUsed, note: note}
		d.pendingEnd = args
		d.pendingTimer = time.AfterFunc(taskEndGraceS*time.Second, func() {
			d.mu.Lock()
			defer d.mu.Unlock()
			d.firePendingLocked()
		})
	}
}

func (d *EventDetector) emitTaskEndLocked(ts float64, taskID *int, totalTokens *int,
	durationS, avgTPS, mtpAcceptance *float64, ctxUsed *int, note string) {
	if taskID != nil {
		if d.endedIDs[*taskID] {
			return // 双源去重
		}
		d.endedIDs[*taskID] = true
		d.endedOrder = append(d.endedOrder, *taskID)
		if len(d.endedOrder) > 200 {
			old := d.endedOrder[0]
			d.endedOrder = d.endedOrder[1:]
			delete(d.endedIDs, old)
		}
	}
	if d.runningID == nil || (taskID != nil && *d.runningID == *taskID) {
		d.taskRunning = false
		d.runningID = nil
	}
	var parts []string
	if taskID != nil {
		parts = append(parts, fmt.Sprintf("任务 #%d 结束", *taskID))
	} else {
		parts = append(parts, "任务结束")
	}
	if note != "" {
		parts = append(parts, "("+note+")")
	}
	var stats []string
	if totalTokens != nil {
		stats = append(stats, fmt.Sprintf("%d tokens", *totalTokens))
	}
	if durationS != nil {
		stats = append(stats, fmtDuration(*durationS))
	}
	if avgTPS != nil {
		stats = append(stats, fmt.Sprintf("平均 %.1f tok/s", *avgTPS))
	}
	if mtpAcceptance != nil {
		stats = append(stats, fmt.Sprintf("MTP 接受率 %.1f%%", *mtpAcceptance*100))
	}
	if ctxUsed != nil {
		stats = append(stats, fmt.Sprintf("上下文 %d", *ctxUsed))
	}
	msg := strings.Join(parts, " ")
	if len(stats) > 0 {
		msg += ": " + strings.Join(stats, " · ")
	}
	d.emitLocked(ts, "info", "task_end", msg)
}

func (d *EventDetector) firePendingLocked() {
	args := d.pendingEnd
	d.pendingEnd = nil
	if d.pendingTimer != nil {
		d.pendingTimer.Stop()
		d.pendingTimer = nil
	}
	if args == nil {
		return
	}
	d.emitTaskEndLocked(args.ts, args.taskID, args.totalTokens, args.durationS,
		args.avgTPS, args.mtpAcceptance, args.ctxUsed, args.note)
}

func (d *EventDetector) cancelPendingLocked() {
	d.pendingEnd = nil
	if d.pendingTimer != nil {
		d.pendingTimer.Stop()
		d.pendingTimer = nil
	}
}

func (d *EventDetector) SetModel(ts float64, modelPath string) {
	if modelPath == "" {
		return
	}
	d.mu.Lock()
	defer d.mu.Unlock()
	if d.modelPath == nil {
		d.modelPath = &modelPath
		return
	}
	if *d.modelPath != modelPath {
		old := *d.modelPath
		d.modelPath = &modelPath
		d.emitLocked(ts, "warn", "model_change", fmt.Sprintf("模型变更: %s → %s", shortPath(old), shortPath(modelPath)))
	}
}

func (d *EventDetector) LlamaBoot(ts float64, info string) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.emitLocked(ts, "info", "llama_boot", "llama 启动: "+info)
}

const alertCooldownS = 30.0

var alertNames = map[string]string{
	"cpu": "CPU", "mem": "内存", "ctx": "上下文", "mtp": "MTP 接受率",
}
var gpuKindNames = map[string]string{
	"util": "利用率", "mem": "显存", "temp": "温度", "power": "功耗",
}

func alertName(metric string) string {
	if n, ok := alertNames[metric]; ok {
		return n
	}
	if strings.HasPrefix(metric, "gpu") && strings.Contains(metric, ".") {
		parts := strings.SplitN(metric, ".", 2)
		return fmt.Sprintf("GPU%s %s", parts[0][3:], gpuKindNames[parts[1]])
	}
	if strings.HasPrefix(metric, "disk:") {
		return "磁盘 " + metric[5:]
	}
	return metric
}

// CheckAlerts 每次快照 evaluate_alerts 后调用：级别变化（升级/恢复）时发事件。
func (d *EventDetector) CheckAlerts(alerts []Alert) {
	now := float64(time.Now().UnixNano()) / 1e9
	d.mu.Lock()
	defer d.mu.Unlock()
	current := make(map[string]Alert)
	for _, a := range alerts {
		if a.Metric != "llama" && a.Metric != "ssh" {
			current[a.Metric] = a
		}
	}
	for metric, a := range current {
		level := a.Level
		prev, ok := d.alertLevels[metric]
		if !ok {
			d.alertLevels[metric] = alertLevel{level: level}
			continue
		}
		if level == prev.level {
			continue
		}
		if prev.lastEmit != nil && now-*prev.lastEmit < alertCooldownS {
			d.alertLevels[metric] = alertLevel{level: level, lastEmit: prev.lastEmit}
			continue
		}
		op := ">="
		if InvertedMetrics[metric] {
			op = "<="
		}
		tag := "警告"
		if level == "danger" {
			tag = "危险"
		}
		msg := fmt.Sprintf("%s %s %s %s（%s）", alertName(metric), fmtAlertVal(metric, a.Value),
			op, fmtAlertVal(metric, a.Threshold), tag)
		d.emitLocked(now, level, "alert", msg)
		d.alertLevels[metric] = alertLevel{level: level, lastEmit: &now}
	}
	for metric, lv := range d.alertLevels {
		if lv.level == "normal" {
			continue
		}
		if _, ok := current[metric]; ok {
			continue
		}
		if lv.lastEmit == nil {
			d.alertLevels[metric] = alertLevel{level: "normal"}
			continue
		}
		if now-*lv.lastEmit < alertCooldownS {
			continue
		}
		d.emitLocked(now, "info", "alert", alertName(metric)+" 恢复正常")
		d.alertLevels[metric] = alertLevel{level: "normal", lastEmit: &now}
	}
}

func (d *EventDetector) List(limit int) []Event {
	d.mu.Lock()
	defer d.mu.Unlock()
	items := make([]Event, len(d.events))
	copy(items, d.events)
	if limit > 0 && len(items) > limit {
		items = items[len(items)-limit:]
	}
	return items
}

func (d *EventDetector) emitLocked(ts float64, level, typ, msg string) {
	d.events = append(d.events, Event{TS: ts, Level: level, Type: typ, Msg: msg})
	if len(d.events) > d.maxlen {
		d.events = d.events[len(d.events)-d.maxlen:]
	}
}

func modelSuffix(name string) string {
	if name == "" {
		return ""
	}
	return " · 模型 " + name
}

func taskStartMsg(taskID, promptTokens *int) string {
	msg := "任务开始"
	if taskID != nil {
		msg = fmt.Sprintf("任务 #%d 开始", *taskID)
	}
	if promptTokens != nil {
		msg += fmt.Sprintf(" (prompt %d tokens)", *promptTokens)
	}
	return msg
}

func shortPath(p string) string {
	if p == "" {
		return p
	}
	return p[strings.LastIndex(p, "/")+1:]
}

func fmtAlertVal(metric string, v float64) string {
	s := fmt.Sprintf("%.1f", v)
	if strings.HasSuffix(metric, ".temp") {
		return s + "°C"
	}
	return s + "%"
}

func fmtDuration(seconds float64) string {
	if seconds < 60 {
		return fmt.Sprintf("%.0fs", seconds)
	}
	m := int(seconds) / 60
	s := int(seconds) % 60
	if m < 60 {
		return fmt.Sprintf("%dm%02ds", m, s)
	}
	h := m / 60
	return fmt.Sprintf("%dh%02dm", h, m%60)
}

File: backend/events.py
"""事件检测器（EventDetector）。

每主机一个状态机，检测状态迁移并产生事件：
- llama: online ⇄ offline
- ssh:   connected ⇄ disconnected
- task:  idle → running (task_start) → idle (task_end, 带统计)
- model: model_path 变化 → model_change
- boot:  日志 "listening"/"model loaded" → llama_boot

事件环形缓冲 200 条，字段 {ts, level, type, msg}。兼容 Python 3.9。
"""
import asyncio
import time
from collections import deque
from typing import Any, Dict, List, Optional

from .config import INVERTED_METRICS


class EventDetector:
    # API 源任务结束事件的宽限期：等日志源补全完整统计（MTP/上下文）
    TASK_END_GRACE_S = 2.0

    def __init__(self, maxlen: int = 200):
        self.events: deque = deque(maxlen=maxlen)
        self._llama_online: Optional[bool] = None
        self._ssh_connected: Optional[bool] = None
        self._task_running: bool = False
        self._running_task_id: Optional[int] = None
        self._model_path: Optional[str] = None
        self._pending_task_end: Optional[tuple] = None
        self._pending_task_end_timer = None
        self._ended_task_ids: deque = deque(maxlen=200)
        self._alert_levels: Dict[str, tuple] = {}

    # ------------------------------------------------------------------
    def emit(self, ts: Optional[float], level: str, type_: str, msg: str, data: Optional[Dict[str, Any]] = None) -> None:
        self.events.append({
            "ts": ts if ts is not None else time.time(),
            "level": level,
            "type": type_,
            "msg": msg,
            "data": data or {},
        })

    def set_llama_online(self, ts: float, online: bool, model_name: str = "") -> None:
        if self._llama_online is None:
            self._llama_online = online
            if online:
                self.emit(ts, "info", "llama_up", "llama 上线" + (" · 模型 %s" % model_name if model_name else ""), data={"model": model_name, "recovered": False})
            return
        if online != self._llama_online:
            self._llama_online = online
            if online:
                self.emit(ts, "info", "llama_up", "llama 恢复在线" + (" · 模型 %s" % model_name if model_name else ""), data={"model": model_name, "recovered": True})
            else:
                self.reset_task_state()
                self.emit(ts, "error", "llama_down", "llama 离线", data={})

    def set_ssh_connected(self, ts: float, connected: bool) -> None:
        if self._ssh_connected is None:
            self._ssh_connected = connected
            return
        if connected != self._ssh_connected:
            self._ssh_connected = connected
            if connected:
                self.emit(ts, "info", "ssh_up", "SSH 重连成功", data={})
            else:
                self.emit(ts, "warn", "ssh_down", "SSH 断开", data={})

    def reset_task_state(self) -> None:
        """重置任务相关状态。llama 重启后任务 ID 从 0 重新计数，
        旧的去重记录会把新任务的结束事件误判为重复而吞掉。"""
        self._task_running = False
        self._running_task_id = None
        self._ended_task_ids.clear()
        self._cancel_pending_task_end()

    def set_task_running(self, ts: float, running: bool, task_id: Optional[int] = None,
                         prompt_tokens: Optional[int] = None) -> None:
        if running and not self._task_running:
            self._task_running = True
            self._running_task_id = task_id
            msg = "任务 #%s 开始" % task_id if task_id is not None else "任务开始"
            if prompt_tokens is not None:
                msg += " (prompt %d tokens)" % prompt_tokens
            self.emit(ts, "info", "task_start", msg, data={"task_id": task_id, "prompt_tokens": prompt_tokens})
        elif (running and self._task_running and task_id is not None
              and task_id != self._running_task_id):
            # 旧任务的结束事件还在宽限期内，新任务已开始：先落盘旧结束事件再开始
            self._fire_pending_task_end()
            self._task_running = True
            self._running_task_id = task_id
            msg = "任务 #%s 开始" % task_id
            if prompt_tokens is not None:
                msg += " (prompt %d tokens)" % prompt_tokens
            self.emit(ts, "info", "task_start", msg, data={"task_id": task_id, "prompt_tokens": prompt_tokens})
        elif not running and self._task_running:
            self._task_running = False
            self._running_task_id = None
            self._cancel_pending_task_end()
            self.emit(ts, "info", "task_end", "任务 #%s 结束" % task_id if task_id is not None else "任务结束", data={"task_id": task_id})

    def task_end_with_stats(self, ts: float, task_id: Optional[int], total_tokens: Optional[int],
                            duration_s: Optional[float], avg_tps: Optional[float],
                            mtp_acceptance: Optional[float] = None,
                            ctx_used: Optional[int] = None,
                            note: Optional[str] = None) -> None:
        # 日志源与 API 源都会调用本方法。日志源统计完整（MTP/上下文），
        # API 源只有 tokens/时长/均速。完整统计立即发；不完整（API 源先到）
        # 延迟 TASK_END_GRACE_S 给日志源补全的机会，超时用部分统计兜底。
        complete = (mtp_acceptance is not None or ctx_used is not None)
        if complete:
            self._cancel_pending_task_end()
            self._emit_task_end(ts, task_id, total_tokens, duration_s, avg_tps,
                                mtp_acceptance, ctx_used, note)
        else:
            if not self._task_running:
                return
            if self._pending_task_end is None:
                self._pending_task_end = (ts, task_id, total_tokens, duration_s,
                                          avg_tps, mtp_acceptance, ctx_used, note)
                try:
                    self._pending_task_end_timer = asyncio.get_running_loop().call_later(
                        self.TASK_END_GRACE_S, self._fire_pending_task_end)
                except RuntimeError:
                    self._fire_pending_task_end()

    def _emit_task_end(self, ts: float, task_id: Optional[int], total_tokens: Optional[int],
                       duration_s: Optional[float], avg_tps: Optional[float],
                       mtp_acceptance: Optional[float], ctx_used: Optional[int],
                       note: Optional[str]) -> None:
        if task_id is not None:
            if task_id in self._ended_task_ids:
                return  # 该任务的结束事件已发过（双源去重）
            self._ended_task_ids.append(task_id)
        if self._running_task_id in (None, task_id):
            self._task_running = False
            self._running_task_id = None
        parts = []
        if task_id is not None:
            parts.append("任务 #%s 结束" % task_id)
        else:
            parts.append("任务结束")
        if note:
            parts.append("(%s)" % note)
        stats = []
        if total_tokens is not None:
            stats.append("%d tokens" % total_tokens)
        if duration_s is not None:
            stats.append(_fmt_duration(duration_s))
        if avg_tps is not None:
            stats.append("平均 %.1f tok/s" % avg_tps)
        if mtp_acceptance is not None:
            stats.append("MTP 接受率 %.1f%%" % (mtp_acceptance * 100))
        if ctx_used is not None:
            stats.append("上下文 %d" % ctx_used)
        msg = " ".join(parts)
        if stats:
            msg += ": " + " · ".join(stats)
        self.emit(ts, "info", "task_end", msg, data={"task_id": task_id, "total_tokens": total_tokens, "duration_s": duration_s, "avg_tps": avg_tps, "mtp_acceptance": mtp_acceptance, "ctx_used": ctx_used, "note": note})

    def _fire_pending_task_end(self) -> None:
        args = self._pending_task_end
        self._pending_task_end = None
        if self._pending_task_end_timer is not None:
            self._pending_task_end_timer.cancel()
            self._pending_task_end_timer = None
        if args is None:
            return
        self._emit_task_end(*args)

    def _cancel_pending_task_end(self) -> None:
        self._pending_task_end = None
        if self._pending_task_end_timer is not None:
            self._pending_task_end_timer.cancel()
            self._pending_task_end_timer = None

    def set_model(self, ts: float, model_path: str) -> None:
        if not model_path:
            return
        if self._model_path is None:
            self._model_path = model_path
            return
        if model_path != self._model_path:
            old = self._model_path
            self._model_path = model_path
            self.emit(ts, "warn", "model_change", "模型变更: %s → %s" % (_short(old), _short(model_path)), data={"old": old, "new": new})

    def llama_boot(self, ts: float, info: str) -> None:
        self.emit(ts, "info", "llama_boot", "llama 启动: " + info, data={"info": info})

    # ------------------------------------------------------------------
    # 阈值穿越事件：按 metric 跟踪级别变化（normal/warn/danger）
    ALERT_COOLDOWN_S = 30.0
    _ALERT_NAMES = {
        "cpu": "CPU",
        "mem": "内存",
        "ctx": "上下文",
        "mtp": "MTP 接受率",
    }
    _GPU_KIND_NAMES = {"util": "利用率", "mem": "显存", "temp": "温度", "power": "功耗"}

    @classmethod
    def _alert_name(cls, metric: str) -> str:
        if metric in cls._ALERT_NAMES:
            return cls._ALERT_NAMES[metric]
        if metric.startswith("gpu") and "." in metric:
            idx, kind = metric.split(".", 1)
            return "GPU%s %s" % (idx[3:], cls._GPU_KIND_NAMES.get(kind, kind))
        if metric.startswith("disk:"):
            return "磁盘 %s" % metric[5:]
        return metric

    def check_alerts(self, alerts: List[Dict[str, Any]]) -> None:
        """每次快照 evaluate_alerts 后调用：级别变化（升级/恢复）时发事件。
        首次观测只记基线不发（基线不占冷却）；同一 metric 两次事件最小间隔
        30s，防阈值边缘抖动（被抑制的变化只更新状态不发事件）。
        llama/ssh 已有专门的上下线事件，不在此重复。"""
        now = time.time()
        current = {a["metric"]: a for a in alerts if a.get("metric") not in ("llama", "ssh")}
        for metric, a in current.items():
            level = a.get("level", "warn")
            prev = self._alert_levels.get(metric)
            if prev is None:
                self._alert_levels[metric] = (level, None)
                continue
            prev_level, last_emit = prev
            if level == prev_level:
                continue
            if last_emit is not None and now - last_emit < self.ALERT_COOLDOWN_S:
                self._alert_levels[metric] = (level, last_emit)
                continue
            op = "<=" if metric in INVERTED_METRICS else ">="
            tag = "危险" if level == "danger" else "警告"
            msg = "%s %s %s %s（%s）" % (
                self._alert_name(metric), _fmt_alert_val(metric, a.get("value")),
                op, _fmt_alert_val(metric, a.get("threshold")), tag)
            self.emit(now, level, "alert", msg, data={"metric": metric, "value": a.get("value"), "op": op, "threshold": a.get("threshold"), "level": level, "recovered": False})
            self._alert_levels[metric] = (level, now)
        for metric, (level, last_emit) in list(self._alert_levels.items()):
            if level == "normal" or metric in current:
                continue
            if last_emit is None:
                # 基线状态（首次观测只记基线、从未发过告警事件）：静默转 normal，
                # 不发"恢复正常"，否则会出现没有对应告警的孤立恢复事件
                self._alert_levels[metric] = ("normal", None)
                continue
            if now - last_emit < self.ALERT_COOLDOWN_S:
                self._alert_levels[metric] = (level, last_emit)
                continue
            self.emit(now, "info", "alert", "%s 恢复正常" % self._alert_name(metric), data={"metric": metric, "value": None, "op": None, "threshold": None, "level": "info", "recovered": True})
            self._alert_levels[metric] = ("normal", now)

    # ------------------------------------------------------------------
    def list(self, limit: int = 50) -> List[dict]:
        items = list(self.events)
        return items[-limit:] if limit > 0 else items

    @property
    def llama_online(self) -> Optional[bool]:
        return self._llama_online

    @property
    def ssh_connected(self) -> Optional[bool]:
        return self._ssh_connected


def _short(path: str) -> str:
    return path.rsplit("/", 1)[-1] if path else path


def _fmt_alert_val(metric: str, v: Any) -> str:
    if v is None:
        return "—"
    s = ("%.1f" % v) if isinstance(v, float) else str(v)
    if metric.endswith(".temp"):
        return s + "°C"
    return s + "%"


def _fmt_duration(seconds: float) -> str:
    if seconds < 60:
        return "%.0fs" % seconds
    m = int(seconds // 60)
    s = int(seconds % 60)
    if m < 60:
        return "%dm%02ds" % (m, s)
    h = m // 60
    return "%dh%02dm" % (h, m % 60)
